pkg/app/ops/insightcollector/collector.go

```diff
@@ -13,3 +13,438 @@
 // limitations under the License.
 
 package insightcollector
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/pipe-cd/pipe/pkg/datastore"
+	"github.com/pipe-cd/pipe/pkg/filestore"
+	"github.com/pipe-cd/pipe/pkg/insightstore"
+	"github.com/pipe-cd/pipe/pkg/model"
+)
+
+var aggregateKinds = []model.InsightMetricsKind{
+	model.InsightMetricsKind_DEPLOYMENT_FREQUENCY,
+	model.InsightMetricsKind_CHANGE_FAILURE_RATE,
+}
+
+// InsightCollector implements the behaviors for the gRPC definitions of InsightCollector.
+type InsightCollector struct {
+	projectStore     datastore.ProjectStore
+	applicationStore datastore.ApplicationStore
+	deploymentStore  datastore.DeploymentStore
+	insightstore     insightstore.Store
+	logger           *zap.Logger
+}
+
+// NewInsightCollector creates a new InsightCollector instance.
+func NewInsightCollector(
+	ds datastore.DataStore,
+	fs filestore.Store,
+	logger *zap.Logger,
+) *InsightCollector {
+	a := &InsightCollector{
+		projectStore:     datastore.NewProjectStore(ds),
+		applicationStore: datastore.NewApplicationStore(ds),
+		deploymentStore:  datastore.NewDeploymentStore(ds),
+		insightstore:     insightstore.NewStore(fs),
+		logger:           logger.Named("insight-collector"),
+	}
+	return a
+}
+
+var (
+	pageSize = 50
+)
+
+func (i *InsightCollector) CollectProjectsInsight(ctx context.Context) error {
+	now := time.Now().UTC()
+	maxCreatedAt := now.Unix()
+
+	for {
+		projects, err := i.projectStore.ListProjects(ctx, datastore.ListOptions{
+			PageSize: pageSize,
+			Filters: []datastore.ListFilter{
+				{
+					Field:    "CreatedAt",
+					Operator: "<",
+					Value:    maxCreatedAt,
+				},
+			},
+			Orders: []datastore.Order{
+				{
+					Field:     "CreatedAt",
+					Direction: datastore.Desc,
+				},
+			},
+		})
+		if err != nil {
+			return err
+		}
+		if len(projects) == 0 {
+			// updated all project's insights completely
+			break
+		}
+
+		for _, p := range projects {
+			for _, k := range aggregateKinds {
+				err = i.updateApplicationChunks(ctx, p.Id, "", k, now)
+				if err != nil {
+					return err
+				}
```

Comment:
I think we should continue working on other projects.

```
if err := i.updateApplicationChunks(ctx, p.Id, "", k, now); err != nil {
    log the error
    continue
}
```

Reply:
Right, thanks :D